src/lib.rs

```diff
@@ -71,10 +71,24 @@ pub trait UdpStack {
 	/// The type returned when we have an error
 	type Error: core::fmt::Debug;
 
-	/// Open a new UDP socket to the given address and port. UDP is connectionless,
-	/// so unlike `TcpStack` no `connect()` is required.
+	/// Open a UDP socket with a dynamically selected port.
+	///
+	/// Selects a port number automatically and opens a UDP socket to the given
+	/// address. UDP is connectionless, so unlike `TcpStack` no `connect()` is
+	/// required.
 	fn open(&self, remote: SocketAddr, mode: Mode) -> Result<Self::UdpSocket, Self::Error>;
 
+	/// Open a new UDP socket with a specified port
+	///
+	/// Opens a new socket with the specified port number to the given address.
+	/// UDP is connectionless, so unlike `TcpStack` no `connect()` is required.
+	fn open_port(
```

Comment:
If we're following specific networking semantics, I think we would actually want to create a bind function with the following signature. It doesn't make sense to have to different open functions.

It's also worth questioning why this would be required in the embedded-nal. Keep in mind that the more functions we add to the trait, the harder it is for users to implement and the less likely people will be to use the `embedded-nal` interface. We should be careful about adding more functions than necessary and over-complicating the trait.

That being said, if we did want this functionality, I believe the proper approach would be to add a `bind` function to bind the socket to a specific port:

```rust
fn bind(&self, socket: Self::UdpSocket, address: SocketAddr) -> Result<Self::UdpSocket, Self::Error>;
```

Reply:
@jonahd-g Would you please provide some sample use cases as to where `bind()`-like functionality would be needed? Is it possible that your use case can be done _outside_ of the NAL?

Reply:
@ryan-summers respectfully, the need to bind to a specific port is such an obviously fundamental requirement of a NAL to me that I don't really understand the question.  Like, if I want a UDP server, I need to be able to listen on a particular port, otherwise nobody knows where to find it.

I'm fine with naming it bind, that makes sense.  And even if you want to go back to the `Option` type arg I'd be okay with that too.

Reply:
No, you're fine. I totally spaced that we can have embedded applications that are servers as well as clients. Bind totally makes sense here. Feel free to ignore my short-sightedness :)

Reply:
> I totally spaced that we can have embedded applications that are servers as well as clients. Bind totally makes sense here. Feel free to ignore my short-sightedness :)

Haha sounds good.

> That being said, I think from an API simplicity level, I'd prefer this being an Option in open(), but I'm noticing that my opinion seems to be shifting back and forth here.

Yes I'm of two minds on this one as well.  The reason I wrote it as a separate method is so that the flow is known at compile time.  If the user only calls the `bind` method, the compiler can omit the `open` method entirely.  `Open` would be pretty much just `make up a port and call bind(port)`.  Although, perhaps the compiler can _already_ statically analyze the usage.  I don't know what kind of optimizations it makes, this could be a moot point.

Reply:
I would personally lean towards following the Berkeley socket API https://en.wikipedia.org/wiki/Berkeley_sockets 

That way we follow some sort of standard, and one that is widely used in C & POSIX, making adoption easier as it looks and feels familiar to devs.

Reply:
@MathiasKoch What would be the list of methods if we did that?

Reply:
Basically yes. 
Where we currently have `open` roughly equivalent to `socket` + `connect` when talking UDP client

Reply:
Ahh, so `connect` would be the antithesis of `bind`.  I like that, I'll make that revision.

Reply:
:+1: